### PR TITLE
Initial stuff for ansible-database-migrations

### PR DESCRIPTION
**Pull request by moshthepitt**
This PR adds the `ansible-database-migrations` role.

The purpose of this role is to build on top of [ansible-sqitch](https://github.com/onaio/ansible-sqitch) and actually run sqitch migrations.  Actually, that is not correct.  This role does not just run sqitch migrations, it does three things:

1. Runs any pre-defined SQL queries that are meant to set up the database.  These may be things such as creating databases and database users, creating postgres extensions etc.
2. Actually running sqitch migrations
3. Setting up certain pre-defined SQL queries to be run periodically by setting them up as cron jobs.  These may be queries such as a daily data dump, refreshing materialized views, etc.

### Other things

- Currently only works for postgres
- Sets up and uses a .pgpass file for authentication
- Periodic SQL query results are output to predefined log files, and logrotate configs are set for these

### Things this role does not do

- This role does not concern itself with installing the database system
- It does not install sqitch directly, that is done through  [ansible-sqitch](https://github.com/onaio/ansible-sqitch)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,35 +1,259 @@
...
 
-A list of other roles hosted on Galaxy should go here, plus any details in regards to parameters that may need to be set for other roles, or variables that are used from other roles.
+Some of the more important variables are briefly described below.  You can see all variables by looking at the [defaults/main.yml](defaults/main.yml) file.
 
-Example Playbook
-----------------
+### database_migrations_system_user
 
-Including an example of how to use your role (for instance, with variables passed in as parameters) is always nice for users too:
+This is the user which is used while running the role.  The default is "nomad" - you know, because they migrate :)
```
Comment (ukanga) [inline README.md:25]: I wonder, should we name the repository and role `ansible-nomad`? Or `ansible-sqitch-migrations`?

File: tasks/setup.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+- name: Copy pgpass credentials
+  template:
+    src: "pgpass.j2"
+    dest: "{{ database_migrations_home }}/.pgpass"
+    owner: "{{ database_migrations_system_user }}"
+    group: "{{ database_migrations_system_group }}"
+    mode: "600"
+  when:
+    - database_migrations_db_connections.values() | length > 0
+    - database_migrations_create_pgpass | bool
```
Comment (ukanga) [inline tasks/setup.yml]: Why is this required given we have `db/setup.yml` below?

File: tasks/clone.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+---
+- name: Clone {{ repo.url }}
+  include_role:
+    name: adamyala.ansible-role-clone-repo
```
Comment (ukanga) [inline tasks/clone.yml:4]: Is there a benefit we get when we use this role instead of the default Ansible [Git module](https://docs.ansible.com/ansible/latest/modules/git_module.html)?

File: meta/main.yml [hunk 4]
```diff
@@ -1,16 +1,27 @@
...
     - name: Ubuntu
       versions:
         - 16.04
         - 18.04
 
-dependencies: []
+dependencies:
+  - role: onaio.sqitch
+    become: true
+    become_user: "root"
```
Comment (ukanga) [inline meta/main.yml]: The default value is `root`, I would exclude this unless the user is someone else other than `root`.

File: README.md (hunk 1 above)
Comment (moshthepitt) [reply README.md:25]: I'll try `ansible-sqitch-migrations` :) - ansible-nomad is tempting but I think the generally accepted way to name ansible roles is in a way that describes what they do

File: tasks/clone.yml (hunk 3 above)
Comment (moshthepitt) [reply tasks/clone.yml:4]: It provides a way to have the git stuff done in a more packaged and re-usable way.  Very similar to what [onaio.ansible-git-clone](https://github.com/onaio/ansible-git-clone) is for.  I am suing this instead of `onaio.ansible-git-clone` because right now onaio.ansible-git-clone is broken for ssh agent forwarding.

File: meta/main.yml (hunk 4 above)
Comment (moshthepitt) [reply meta/main.yml]: 100% will remove this.

File: tasks/setup.yml (hunk 2 above)
Comment (moshthepitt) [reply tasks/setup.yml]: @ukanga I was thinking that one may want to run/rerun this role and not touch any existing .pgpass giles

File: README.md (hunk 1 above)
Comment (moshthepitt) [reply README.md:25]: Done

File: README.md [hunk 5]
```diff
@@ -1,35 +1,259 @@
...
+- **registry**: (optional) the sqitch registry to use, default is "sqitch"
+- **extra_args**: (optional) extra args for sqitch deploy command, default is "--verify"
+- **engine**: (optional) database engine to use, default is pg
+
+### sqitch_migrations_sql_jobs
+
+An object containing SQL queries that should be executed periodically.
+
+```yml
+sqitch_migrations_sql_jobs:
```
Comment (gstuder-ona) [inline README.md:125]: yay!

IMO we should move to as db-centered an approach as we can, so we can define mat views alongside their refresh intervals (something like a system.cron table or pg_cron) vs requiring redeploy, but we can then just use this to run a SP or something to check dynamically what needs to happen in the next time slice.

File: README.md [hunk 6]
```diff
@@ -1,35 +1,259 @@
...
+              host: localhost
+              port: 5432
+              database: productiondb
+        - sqitch_migrations_sql_for_setup:
+            db_admin:
+              - "{{ sqitch_migrations_home }}/data-solutions/OpenSRP/2-reveal/setup/extensions.sql"
+        - sqitch_migrations_sqitch_plans:
+            db_admin:
+              - directory: "{{ sqitch_migrations_home }}/data-solutions/examples/sqitch_test"
+            db_user:
```
Comment (gstuder-ona) [inline README.md:188]: This one is a little scary to me - if we start assembling a lot of micro-migrations together in playbooks vs data-solutions repo, we won't be able to test stuff very well without a lot of infrastructure.

Do we want to push more of this down to data solutions - you can definitely use subdirectories and symlinks, for example, for shared migration piece.  Alternately copy-paste, if each project is really a separate fork at the time-of-creation, could make sense in this case.

EDIT: On second glance I'd say it's more a usage thing, in that we can decide what granularity to break up our scripts but it's still useful here to be able to specify multiple migrations

**Review by gstuder-ona — COMMENTED** [review]
Looks good, more of a question - would it make sense to by-default include ansible-sqitch as a dependency (with a flag to turn that off)?  I know for git-clone, for example, which is similar, it's annoying to always forget to install git first and Stouts.backup doesn't make you install duplicity separately.

**Review by gstuder-ona — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (moshthepitt) [reply README.md:125]: Yes I agree.  We should push as much as possible to the DB itself.  However I think when we are using Amazon's RDS our hands are a little tied because they don't allow pg_cron.  Anyway, I think we are generally on the same page and should watch out that this is not abused.

File: README.md (hunk 6 above)
Comment (moshthepitt) [reply README.md:188]: I've actually been really torn about how the Reveal migrations are currently structured.  I might change it.  Maybe.  Anyway, I think sqitch was meant to be used more like what you describe here:

> Do we want to push more of this down to data solutions - you can definitely use subdirectories and symlinks, for example, for shared migration piece. Alternately copy-paste, if each project is really a separate fork at the time-of-creation, could make sense in this case.

So I do agree that we should be wary of using this role to orchestrate migrations.  We should let sqitch do that.

Though as noted in your edit we should let it be possible to run multiple migrations e.g. in the case of discover.ona.io we may have migrations for more than one projects living in the same ansible inventory.

**Comment by moshthepitt** [comment]
> Looks good, more of a question - would it make sense to by-default include ansible-sqitch as a dependency (with a flag to turn that off)? I know for git-clone, for example, which is similar, it's annoying to always forget to install git first and Stouts.backup doesn't make you install duplicity separately

It is set up as a dependency.  If you don't have sqitch running then this role will install it ready for use with postgres.
